src/common/iceberg.cpp

```diff
@@ -3,90 +3,301 @@
 	if (iceberg_format_version == 1) {
-		schema = avro::compileJsonSchemaFromString(MANIFEST_SCHEMA_V1);
-		avro::DataFileReader<c::manifest_file_v1> dfr(std::move(stream), schema);
-		c::manifest_file_v1 manifest_list;
-		while (dfr.read(manifest_list)) {
-			ret.emplace_back(IcebergManifest(manifest_list));
-		}
+		produce_manifests = [](DataChunk &input, case_insensitive_map_t<ColumnIndex> &name_to_vec, vector<IcebergManifest> &result) {
+			auto manifest_paths = FlatVector::GetData<string_t>(input.data[0]);
+
+			for (idx_t i = 0; i < input.size(); i++) {
+				IcebergManifest manifest;
+				manifest.manifest_path = manifest_paths[i].GetString();
+				manifest.content = IcebergManifestContentType::DATA;
+				manifest.sequence_number = 0;
+
+				result.push_back(manifest);
+			}
+		};
+	} else if (iceberg_format_version == 2) {
+		produce_manifests = [](DataChunk &input, case_insensitive_map_t<ColumnIndex> &name_to_vec, vector<IcebergManifest> &result) {
+			auto manifest_paths = FlatVector::GetData<string_t>(input.data[0]);
+			auto content = FlatVector::GetData<int32_t>(input.data[3]);
+			auto sequence_numbers = FlatVector::GetData<int64_t>(input.data[4]);
+
+			for (idx_t i = 0; i < input.size(); i++) {
+				IcebergManifest manifest;
+				manifest.manifest_path = manifest_paths[i].GetString();
+				manifest.content = IcebergManifestContentType(content[i]);
+				manifest.sequence_number = sequence_numbers[i];
+
+				result.push_back(manifest);
+			}
+		};
 	} else {
-		schema = avro::compileJsonSchemaFromString(MANIFEST_SCHEMA);
-		avro::DataFileReader<c::manifest_file> dfr(std::move(stream), schema);
-		c::manifest_file manifest_list;
-		while (dfr.read(manifest_list)) {
-			ret.emplace_back(IcebergManifest(manifest_list));
-		}
+		throw InvalidInputException("iceberg_format_version %d not handled", iceberg_format_version);
 	}
 
+	do {
+		TableFunctionInput function_input(bind_data.get(), nullptr, global_state.get());
+		result.Reset();
+		avro_scan.function(context, function_input, result);
+
+		idx_t count = result.size();
+		for (auto &vec : result.data) {
+			vec.Flatten(count);
+		}
+
+		produce_manifests(result, name_to_vec, ret);
+	} while (result.size() != 0);
 	return ret;
 }
 
-vector<IcebergManifestEntry> IcebergTable::ReadManifestEntries(const string &path, FileSystem &fs,
+static bool VerifyManifestEntrySchema(case_insensitive_map_t<ColumnIndex> &name_to_vec, idx_t iceberg_format_version) {
+	if (!name_to_vec.count("status")) {
+		return false;
+	}
+	if (!name_to_vec.count("file_path")) {
+		return false;
+	}
+	if (!name_to_vec.count("file_format")) {
+		return false;
+	}
+	if (!name_to_vec.count("record_count")) {
+		return false;
+	}
+	if (iceberg_format_version >= 2 && !name_to_vec.count("content")) {
+		return false;
+	}
+	return true;
+}
+
+vector<IcebergManifestEntry> IcebergTable::ReadManifestEntries(ClientContext &context, const string &path,
```

Comment:
FIXME: we can probably create a single scan for all manifest entries, which can be executed lazily

Reply:
It says outdated because I refactored this method to deduplicate the `read_avro` scan boilerplate